### ISSUE TITLE
Issue with ctors/dtors

### ISSUE THREAD

**Issue by avncharlie** [GrammaTech/gtirb-pprinter#17]
I was attempting to disassemble and pretty-print the `ls` program on my Ubuntu 22 system. I've attached it here: [ls.zip](https://github.com/user-attachments/files/18248398/ls.zip)

The pretty-printed binary however, segfaults. This appears to be the issue:
- The original binary contains a .ctors and .dtors section, both containing: `0x0`, `0xffffffff`
- These sections correctly appear in the generated assembly
- Upon re-assembly / linking, these sections get (automatically) converted to .init_array and .fini_array sections
- Upon startup, the linker attempts to run the first item in the .init_array as a startup function, which is 0x0 and causes a segfault. Here is the relevant linker code: https://github.com/lattera/glibc/blob/895ef79e04a953cac1493863bcae29ad85657ee1/elf/dl-init.c#L72

To fix it, I can manually delete the bytes (but not the labels) in the ctors/dtors section and the init_array and fini_arrays end up having 0 size, and the binary runs fine.

I'm unsure why my original binary doesn't segfault itself with the bogus values in its ctors and dtors, but appears to segfault on conversion to .init_array and .fini_array. Strangely, my version of `ls` is different to the `ls` in the ubuntu:22.04 docker image, despite both having the same version string (ls (GNU coreutils) 8.32). The docker version doesn't have .ctors/.dtors, but has .init_array/.fini_array.

Would you know what causes this issue on pretty printing / linking?

**Comment by junghee** [comment]
Hi @avncharlie 
 
Thank you for bringing up the issue.
 
As you observed, the assembler automatically converts .ctors and .dtors into .init_array and .fini_array, respectively.
At runtime, the loader initially looks for the .init_array section and calls each entry within it.
This section is expected to contain a sequence of valid address pointers.
However, in the original binary, .ctors contains -1 followed by 0.
In loading, referencing the first entry, address -1, leads to a segfault.
 
The key takeaway is that gtirb-pprinter outputs the assembly code reflecting the original binary structure, including the legacy .ctors and .dtors sections.
However, the assembler automatically converts .ctors and .dtors into .init_array and .fini_array because modern toolchains follow the ELF specification's recommendation to use .init_array and .fini_array for initialization and finalization routines. The .ctors and .dtors sections are considered legacy and are deprecated in favor of .init_array and .fini_array.
This behavior of conversion was unexpected on our part.
 
To address the issue, your workaround of commenting out the invalid entries in .ctors and .dtors is a viable solution. Another potential workaround is to rename .ctors and .dtors to prevent the assembler from converting them into .init_array and .fini_array.

**Comment by avncharlie** [comment]
Hello,
Thanks for the response.

> invalid entries in .ctors and .dtors

Whats confusing is that I don't believe that these entries are invalid, as the original binary runs without segfaulting. To test for yourself, the `ls` binary in the ubuntu:12.04 docker image uses ctors/dtors, both also containing [-1, 0], and also segfaulting when rewritten as these sections turn into init_array/fini_array.

Maybe there is a difference in how the ctors/dtors sections are handled compared to init_array/fini_array? A guess I had is that maybe ctors/dtors indicate end of array with a 0 or -1, which is incorrectly interpreted as a constructor/destructor function when placed in the init_array/fini_array section?

**Comment by junghee** [comment]
The original `ls` binary does not crash because it uses `.ctors` and `.dtors` sections, which are not processed by the loader in the same way as `.init_array` and `.fini_array`: i.e., the loader explicitly looks for `.init_array` and `.fini_array` and calls each entry within those sections, but this binary does not have them, so it does not crash.

In the original binary, the contents -1 (at 0x21000) and 0 in the `.ctors` section are handled by the function `FUN_170f0` in the binary, as shown below:
```
#-----------------------------------
.type FUN_170f0, @function
#-----------------------------------
FUN_170f0:
    endbr64
    movq .L_21000(%rip), %rax
    cmpq $-1, %rax
    je .L_17130

    pushq %rbp
    movq %rsp, %rbp
    pushq %rbx
    leaq .L_21000(%rip), %rbx
    subq $8, %rsp
    nop
    .align 8
.L_17118:
    callq *%rax
    movq -8(%rbx), %rax
    subq $8, %rbx
    cmpq $-1, %rax
    jne .L_17118

    movq -8(%rbp), %rbx
    leave
    retq
.align 16
.L_17130:
    retq
.size FUN_170f0, . - FUN_170f0
```
In this function, the value -1 serves as a terminator for the loop. This logic ensures the entries in the `.ctors` section are processed correctly without causing a segfault.

The rewritten `ls` binary, however, segfaults because the assembler automatically converts `.ctors` and `.dtors` into `.init_array` and `.fini_array`. The loader then process `.init_array` and `.fini_array` by calling each entry directly, interpreting -1 as an addresses and attempting to dereference it. This behavior is unexpected and causes the segfault.

The root cause is that the original `ls` binary was built using an older toolchain, which uses `.ctors` and `.dtors` with their specific semantics.
In contrast, the rewritten binary is reassembled with a newer assembler and executed using a newer runtime toolchain, which operates under different assumptions about section usage and processing.

Hope this answers your question.

**Comment by avncharlie** [comment]
Right, so ctors/dtors are processed within the binary (not the dynamic linker), and the function that processes them is called by the _init function in the .init section. It feels like an oversight from the linker to blindly convert these sections to .init_array/.fini_array when they don't have the same semantics. 

I think the best fix is to rename the sections, as this prevents converting them into the special init_array/fini_array sections, but also keeps their original functionality, as _init function will still be called on startup and execute the pointers within these sections as the original binary would.

I think this solves my problem, so closing the issue now.
Thanks for the help!
